### ISSUE TITLE
Strange errors message for invalid input to the new boxplot and violinplot

### ISSUE THREAD

**Issue by shoyer** [mwaskom/seaborn#445]
I'm running the dev version of Seaborn to test out the boxplot/violinplot refactor.

I think this functions could use more input validation -- they aren't giving sensible error message when the `y` or `hue` arguments are invalid. For example:
```python
tips = sns.load_dataset("tips")
sns.boxplot("day", "tip", "sexy", tips)
```
```
KeyError: 'Thur'
```

I'm also seeing some even stranger error messages that I think have the same source, though I haven't been able to reproduce them using a dataset I can share:
```python
sns.violinplot('x', 'not a real var', data=df.dropna())
```
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-169-2fbc318d67a2> in <module>()
----> 1 sns.boxplot('x', 'not a real var', data=df.dropna())

/Users/shoyer/dev/seaborn/seaborn/distributions.pyc in boxplot(x, y, hue, data, order, hue_order, orient, color, palette, saturation, width, fliersize, linewidth, whis, notch, ax, **kwargs)
   1161     plotter = _BoxPlotter(x, y, hue, data, order, hue_order,
   1162                           orient, color, palette, saturation,
-> 1163                           width, fliersize, linewidth)
   1164 
   1165     if ax is None:

/Users/shoyer/dev/seaborn/seaborn/distributions.pyc in __init__(self, x, y, hue, data, order, hue_order, orient, color, palette, saturation, width, fliersize, linewidth)
     30                  width, fliersize, linewidth):
     31 
---> 32         self.establish_variables(x, y, hue, data, orient, order, hue_order)
     33         self.establish_colors(color, palette, saturation)
     34 

/Users/shoyer/dev/seaborn/seaborn/distributions.pyc in establish_variables(self, x, y, hue, data, orient, order, hue_order)
    202 
    203                 # Group the numeric data
--> 204                 grouped_vals = vals.groupby(groups)
    205                 plot_data = [grouped_vals.get_group(g) for g in order]
    206                 plot_data = [d.values for d in plot_data]

/Users/shoyer/dev/pandas/pandas/core/generic.pyc in groupby(self, by, axis, level, as_index, sort, group_keys, squeeze)
   2892         axis = self._get_axis_number(axis)
   2893         return groupby(self, by=by, axis=axis, level=level, as_index=as_index,
-> 2894                        sort=sort, group_keys=group_keys, squeeze=squeeze)
   2895 
   2896     def asfreq(self, freq, method=None, how=None, normalize=False):

/Users/shoyer/dev/pandas/pandas/core/groupby.pyc in groupby(obj, by, **kwds)
   1191         raise TypeError('invalid type: %s' % type(obj))
   1192 
-> 1193     return klass(obj, by, **kwds)
   1194 
   1195 

/Users/shoyer/dev/pandas/pandas/core/groupby.pyc in __init__(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, squeeze)
    381         if grouper is None:
    382             grouper, exclusions, obj = _get_grouper(obj, keys, axis=axis,
--> 383                                                     level=level, sort=sort)
    384 
    385         self.obj = obj

/Users/shoyer/dev/pandas/pandas/core/groupby.pyc in _get_grouper(obj, key, axis, level, sort)
   2139 
   2140         ping = Grouping(group_axis, gpr, obj=obj, name=name,
-> 2141                         level=level, sort=sort, in_axis=in_axis)
   2142 
   2143         groupings.append(ping)

/Users/shoyer/dev/pandas/pandas/core/groupby.pyc in __init__(self, index, grouper, obj, name, level, sort, in_axis)
   1868         self.name = name
   1869         self.level = level
-> 1870         self.grouper = _convert_grouper(index, grouper)
   1871         self.index = index
   1872         self.sort = sort

/Users/shoyer/dev/pandas/pandas/core/groupby.pyc in _convert_grouper(axis, grouper)
   2163             return grouper.values
   2164         else:
-> 2165             return grouper.reindex(axis).values
   2166     elif isinstance(grouper, (list, Series, Index, np.ndarray)):
   2167         if len(grouper) != len(axis):

/Users/shoyer/dev/pandas/pandas/core/series.pyc in reindex(self, index, **kwargs)
   2148     @Appender(generic._shared_docs['reindex'] % _shared_doc_kwargs)
   2149     def reindex(self, index=None, **kwargs):
-> 2150         return super(Series, self).reindex(index=index, **kwargs)
   2151 
   2152     def reindex_axis(self, labels, axis=0, **kwargs):

/Users/shoyer/dev/pandas/pandas/core/generic.pyc in reindex(self, *args, **kwargs)
   1729         # perform the reindex on the axes
   1730         return self._reindex_axes(axes, level, limit,
-> 1731                                   method, fill_value, copy).__finalize__(self)
   1732 
   1733     def _reindex_axes(self, axes, level, limit, method, fill_value, copy):

/Users/shoyer/dev/pandas/pandas/core/generic.pyc in _reindex_axes(self, axes, level, limit, method, fill_value, copy)
   1741             ax = self._get_axis(a)
   1742             new_index, indexer = ax.reindex(
-> 1743                 labels, level=level, limit=limit, method=method)
   1744 
   1745             axis = self._get_axis_number(a)

/Users/shoyer/dev/pandas/pandas/core/index.py in reindex(self, target, method, level, limit)
   4141             else:
   4142                 if self.is_unique:
-> 4143                     indexer = self.get_indexer(target, method=method,
   4144                                                limit=limit)
   4145                 else:

/Users/shoyer/dev/pandas/pandas/core/index.py in from_tuples(cls, tuples, sortorder, names)
   3654         --------
   3655         MultiIndex.from_arrays : Convert list of arrays to MultiIndex
-> 3656         MultiIndex.from_product : Make a MultiIndex from cartesian product
   3657                                   of iterables
   3658         """

/Users/shoyer/dev/pandas/pandas/lib.so in pandas.lib.tuples_to_object_array (pandas/lib.c:57572)()

ValueError: Buffer dtype mismatch, expected 'Python object' but got 'long'
```

**Comment by mwaskom** [comment]
Hm, good catch, though this is somewhat a consequence of trying to be as ecumenical as possible in the inputs we accept. i.e., the value can be either the name of a column in `data`, or an array or series of values (even when you also passed `data`). And IIRC, testing whether an array and/or series is "in" a Dataframe raises an exception? So that validation code will be a little messy. 

But, yeah, could be confusing when people typo....

FYI this code is currently wrapped up in a separate branch so a small fix here won't merge cleanly.

**Comment by mwaskom** [comment]
Closed #445 via #466
